### ISSUE TITLE
Missing Oracle Support

### ISSUE THREAD

**Issue by jtysper** [dnvriend/akka-persistence-jdbc#59]
Hi,

after being back from vacation, I'd like to continue our discussion about Oracle support.
I originally encountered issues consuming events from read-journals.

After receiving eventsByTag the EventEnvelops don't contain the expected data.
The "offset" value of the last consumed event is stored and used for later queries but contains an invalid value. Somehow I still don't get around [JdbcReadJournal.scala#L116](https://github.com/dnvriend/akka-persistence-jdbc/blob/master/src/main/scala/akka/persistence/jdbc/query/scaladsl/JdbcReadJournal.scala#L116).
It seems to me that the first parameter should not be the ordering but the total offset of the row within the query (offset parameter + the position within result) like it was before 0d02096a79feeb0180efc498ba67b136afe56bec.

I don't really understand, why this isn't an issue in other DBs, though.
Could you please share your thoughts on this?

I'd like to help if I can but still have way to many questions...

Regards
Jörn

**Comment by dnvriend** [comment]
Hi, welcome back. You have some very interesting questions, and one of them is the ordering/offset. 

The old implementation just read events and numbered them with a zipWithIndex operation. There was no explicit ordering because the CREATED field is a timestamp and with a bulk-insert these events would all have the same value. Because of this, there was no ordering of events in the journal.

By adding the ordering field, a total ordering of events can be created, something that Martin Zapletal kinda touches in his talk [Data in Motion: Streaming Static Data Efficiently in Akka Persistence](https://www.youtube.com/watch?v=K4FY0XKediU). gives us the opportunity to use the ordering field as the `the-offset-to-store` for resumable queries to, for example, create projections. This works, I use this strategy.

Introducing a sequential offset for the eventsByTag query, would introduce a `drop` stage, which would cause a sequential scan of the journal for each query up to the offset. This could become a problem because the offset is a `Long` value. It would also introduce a derivative, the offset, which would be a number, starting from 1 which would increment with one. For example: List[(ordering: Long, event: Any, offset: Long)] = List((303, Foo, 1), (311, Foo, 2), (400, Foo, 3)) and so on. The values are fictive of course. There would be use cases where this could be a nice feature to have, when you wish to have event with offset 10 from this stream. This would cause the stream not to emit events, until there are more than 10 events. But for creating projections, the current implementation would not create an issue and has no performance penalty. 

Maybe the offset should be sequential and starting from 1, incrementing by one, with every materialization, I'm not sure. I'm wondering about performance penalty though for it needs an (unnecessary) journal scan. 

Good questions, and material for further discussion :)

Regards,

Dennis

**Comment by jtysper** [comment]
Hi Dennis,

thank you very much for the exhaustive explanation! 
Please have a look at the eventsByTag method of OracleReadJournalDao.
I would be fine with using the absolute order as offset value. But your restricting the query to rownum instead.
In your above example the last offset would be 400 or 3 - depending on using the order or offset.
When you use the order, the query will drop the first 399 lines though, which I assume is not intended.

Regards
Jörn

**Comment by dnvriend** [comment]
Hi, as stated in the changelog (2.6.0), there is no support for Oracle (yet) as the implementation has changed and the Oracle implementation broke as a cause of that. The Oracle query(ies) are one part of the refactoring that should be updated. I took a (very) short look at Oracle 12 and that DB maybe has support for the SERIAL data type. Also there is a docker image for [Oracle 12C](https://hub.docker.com/r/sath89/oracle-12c/) that would be great for testing. There is also a branch [dnvriend/ORACLE_12C](https://github.com/dnvriend/akka-persistence-jdbc/tree/dnvriend/ORACLE_12C) that you could use to take a look at if you could fix it. That branch is based on v2.6.0, maybe merge v2.6.1 first and work from there.

**Comment by jtysper** [comment]
Hi,

I tried a few things out.
What did the trick was that the sequence had to be reset after each test (which unfortunately isn't exactly elegant in Oracle, so I put it into a procedure).
The other thing was - as I mentioned - that the queryByTag had to compare against the order column and not the rownum.

Tomorrow I have to clean up some things and rebase with 2.6.3.

I also changed some other minor things feel free to have a look on it already at  https://github.com/jtysper/akka-persistence-jdbc/tree/ORACLE_12C

Regards
Jörn
